src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,319 @@
+    // Like ArrayBuffer, this manages a sliding buffer where bytes can be added at the end and removed at the beginning.
+    // Unlike ArrayBuffer, the buffer itself is managed using 16K blocks which are added/removed to the block list as necessary.
+
+    // [ActiveBuffer] contains the current buffer contents; these bytes will be preserved on any call to TryEnsureAvailableBytesUpToLimit.
+    // [AvailableBuffer] contains the available bytes past the end of the current content,
+    // and can be written to in order to add data to the end of the buffer.
+    // Commit(byteCount) will extend the ActiveBuffer by [byteCount] bytes into the AvailableBuffer.
+    // Discard(byteCount) will discard [byteCount] bytes as the beginning of the ActiveBuffer.
+    // TryEnsureAvailableBytesUpToLimit will grow the buffer if necessary; *however*, this may invalidate
+    // old values of [ActiveBuffer] and [AvailableBuffer], so they must be retrieved again.
+
+    internal struct MultiArrayBuffer : IDisposable
+    {
+        private byte[]?[]? _blocks;
+        private int _blockCount;
+        private int _activeStart;
+        private int _availableStart;
+
+        // Invariants:
+        // 0 <= _activeStart <= _availableStart <= total buffer size (i.e. _blockCount * BlockSize)
+
+        private const int BlockSize = 16 * 1024;
+
+        public MultiArrayBuffer(int initialinitialBufferSize)
+        {
+            // [initialBufferSize] is ignored for now;
+            // I kept it because some callers are passing useful info here that we might want to act on in the future.
+
+            _blocks = null;
+            _blockCount = 0;
+            _activeStart = 0;
+            _availableStart = 0;
+        }
+
+        public void Dispose()
+        {
+            _activeStart = 0;
+            _availableStart = 0;
+
+            if (_blocks is not null)
+            {
+                for (int i = 0; i < _blocks.Length; i++)
+                {
+                    if (_blocks[i] is not null)
+                    {
+                        ArrayPool<byte>.Shared.Return(_blocks[i]!);
+                        _blocks[i] = null;
+                    }
+                }
+
+                _blocks = null;
+            }
+        }
+
+        public MultiMemory ActiveMemory => _blocks is null ? MultiMemory.Empty : new MultiMemory(_blocks!, _activeStart, _availableStart - _activeStart);
+
+        public MultiMemory AvailableMemory => _blocks is null ? MultiMemory.Empty : new MultiMemory(_blocks!, _availableStart, _blockCount * BlockSize - _availableStart);
+
+        public void Discard(int byteCount)
+        {
+            Debug.Assert(byteCount >= 0);
+            Debug.Assert(byteCount <= ActiveMemory.Length, $"MultiArrayBuffer.Discard: Expected byteCount={byteCount} <= {ActiveMemory.Length}");
+
+            int oldStartBlock = _activeStart / BlockSize;
```

Comment:
JIT produces optimized division for this. Casting to `uint`, then the JIT emits bit-shifts.
```suggestion
            int oldStartBlock = (int)((uint)_activeStart / BlockSize);
```
But code looks uglier.

The same on other places where `/ BlockSize`.

Reply:
Yeah, I forgot that it will only optimize to a bit-shift if it's uint. I should probably change the fields here to be uint.